flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SharedSlot.java

```diff
@@ -221,7 +227,14 @@ public void release(Throwable cause) {
 			logicalSlotFuture.thenAccept(logicalSlot -> logicalSlot.release(cause));
 		}
 		requestedLogicalSlots.clear();
-		releaseCallback.accept(executionSlotSharingGroup);
+		releaseExternally();
+	}
+
+	private void releaseExternally() {
+		if (state == State.ALLOCATED && requestedLogicalSlots.values().isEmpty()) {
+			externalReleaseCallback.accept(executionSlotSharingGroup);
+			state = State.RELEASED;
```

Comment:
The state needs to be updated before calling `externalReleaseCallback`.
This is because `SharedSlot#release()` can be triggered in the calling stack.